File: mechanics/src/mechanics/mass_spring_solver.py
# -*- python -*-
# -*- coding: latin-1 -*-
#
#       MassSpring : mechanics package
#
#       Copyright or © or Copr. 2006 INRIA - CIRAD - INRA  
#
#       File author(s): Jerome Chopard <jerome.chopard@sophia.inria.fr>
#
#       Distributed under the Cecill-C License.
#       See accompanying file LICENSE.txt or copy at
#           http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html
# 
#       VPlants WebSite : https://gforge.inria.fr/projects/vplants/
#

__doc__ = """
This module provide an interface for deformation of spring systems
"""

__license__= "Cecill-C"
__revision__=" $Id: mass_spring_solver.py 9030 2010-06-01 09:00:10Z chopard $ "

from numpy import array,zeros
from numpy.linalg import norm

LARGE = 1e9

class DivergenceError (Exception) :
	"""Error raised if an algorithm do not converge
	"""
	pass

class MSSolver (object) :
	"""Integration method for mass spring systems
	
	basic implementation of mass spring resolution using a forward Euler method.
	"""
	
	dim = None
	"""dimension of space
	"""
	
	def __init__ (self, particules, actor_list, external_constraint,
	                    logout = None) :
		"""Constructor
		
		Initialise integrator.
		
		:Parameters:
		 - `particles` (dict of (pid|Particule) ) - set of particules
		    that will be connected by springs
		 - `actor_list` (list of :class:`Actor`) - interactions between points
		                         (springs, dampers, ...)
		 - `external_constraint` (function) - boundary function taking
		                                      this solver as argument
		 - `logout` (outstream) - an outfile stream like structure
		                          used to write some debug informations.
		                          If None, no output messages are printed
		"""
		self._particules = particules
		self._actor_list = actor_list
		self._external_constraint = external_constraint
		self._force = {}
		self.logout = logout
	
	########################################
	#
	#		accessors
	#
	########################################
	def position (self, pid) :
		"""Retrieve position of a given point
		
		:Parameters:
		 - `pid` (pid) - id of the point
		
		:Returns Type: Vector
		"""
		return self._particules[pid].position()
	
	def set_position (self, pid, vec) :
		"""Set the position of a given point
		
		:Parameters:
		 - `pid` (pid) - id of the point
		 - `vec` (Vector) - new geometrical position of the point
		"""
		self._particules[pid].set_position(array(vec) )
	
	def force (self, pid) :
		"""Retrieve the force applied on a point
		
		:Parameters:
		 - `pid` (pid) - id of the point
		
		:Returns Type: Vector
		"""
		return self._force[pid]
	
	def set_force (self, pid, F) :
		"""Apply a force on a given point
		
		:Parameters:
		 - `pid` (pid) - id of the point
		 - `F` (Vector) - new force applied on the point
		"""
		self._force[pid] = array(F)
	
	def force_max (self) :
		"""Return the biggest force applied in the system
		
		Use forces computed with :func:`compute_force`
		
		:Returns Type: float
		"""
		return max(norm(F) for F in self._force.itervalues() )
	
	def speed_max (self) :
		"""Return the speed of the fastest point in the system
		
		.. warning:: Uses velocities computed during the last time step
		             and stored into particules
		
		:Returns Type: float
		"""
		return max(norm(p.velocity() ) for p in self.particules.itervalues() )
	
	########################################
	#
	#		compute deformation
	#
	########################################
	def compute_forces (self) :
		"""Compute forces acting on points
		
		Use previously stored positions
		
		:Returns: None, modify forces stored inside the algo struct.
		"""
		#compute forces
		for pid in self._particules :
			self._force[pid] = zeros(self.dim)
		for actor in self._actor_list :
			actor.assign_forces(self._force,self._particules)
		
		#external constraints
		self._external_constraint(self)
	
	def deform (self, dt) :
		"""Compute deformation of the system
		
		Modify particules parameters internally
		
		:Parameters:
		 - `dt` (float) - time step hint
		
		:Returns: the actual dt used for computation
		
		:Returns Type: float
		"""
		raise NotImplementedError
	
	def deform_to_equilibrium (self, dt, force_threshold) :
		"""Compute equilibrium positions of the system
		
		While forces are bigger than a given threshold, call :func:`deform`
		
		:Parameters:
		 - `dt` (float) - time step
		 - `force_threshold` (float) - small force below which the system is
		                               considered to be in equilibrium
		"""
		Fmax = 2 * force_threshold
		while Fmax > force_threshold :
			dt = self.deform(dt)
			Fmax = self.force_max()
			if self.logout is not None :
				self.logout.write("Fmax: %e\n" % Fmax)
				self.logout.flush()


########################################################################
#
#	static solver used to compute forces
#
########################################################################
class StaticSolver2D (MSSolver) :
	"""Compute force generated by a set of actors for a given configuration
	"""
	dim = 2

class StaticSolver3D (MSSolver) :
	"""Compute force generated by a set of actors for a given configuration
	"""
	dim = 3
########################################################################
#
#	simple forward euler
#
########################################################################
class ForwardEuler (MSSolver) :
	"""Compute evolution using a simple Euler integration scheme
	"""
	
	def deform (self, dt) :
		"""Compute deformation of the system
		
		actually call compute_forces then update internally stored particules
		
		:Parameters:
		 - `dt` (float) - time step
		
		:Return: the actual dt used for computation
		
		:Returns Type: float
		"""
		#compute forces
		self.compute_forces()
		
		#compute new positions
		force = self._force
		
		for pid,p in self._particules.iteritems() :
			velocity = p.velocity() + force[pid] * (dt / p.weight() )
			pos = p.position() + velocity * dt
			
			p.set_velocity(velocity)
			p.set_position(pos)
		
		#return
		return dt# * self._nb_steps


class ForwardEuler2D (ForwardEuler) :
	dim = 2


class ForwardEuler3D (ForwardEuler) :
	dim = 3

########################################################################
#
#	Runge et Kuta order 4
#
########################################################################
class RungeKutta (MSSolver) :
	"""Compute equilibrium using a Runge et Kutta integration scheme.
	"""
	
	def deform (self, dt) :
		"""Deform positions using RG4 scheme.
		
		.. seealso:: `http://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods`
		
		0   |
		1/2 | 1/2
		1/2 | 0   1/2
		1   | 0   0   1
		---------------------
		    | 1/6 1/3 1/3 1/6
		"""
		pts = self._particules
		
		#init
		x0 = dict( (pid,p.position() ) for pid,p in pts.iteritems() )
		v0 = dict( (pid,p.velocity() ) for pid,p in pts.iteritems() )
		
		#K1
		self.compute_forces()
		dx1 = dict( (pid,vec * dt) \
		             for pid,vec in v0.iteritems() )
		dv1 = dict( (pid,vec * (dt / pts[pid].weight() ) ) \
		             for pid,vec in self._force.iteritems() )
		
		#K2
		for pid,p in pts.iteritems() :
			p.set_position(x0[pid] + dx1[pid] / 2.)
			p.set_velocity(v0[pid] + dv1[pid] / 2.)
		self.compute_forces()
		dx2 = dict( (pid,(vec + dv1[pid] / 2.) * dt) \
		             for pid,vec in v0.iteritems() )
		dv2 = dict( (pid,vec * (dt / pts[pid].weight() ) ) \
		             for pid,vec in self._force.iteritems() )
		
		#K3
		for pid,p in pts.iteritems() :
			p.set_position(x0[pid] + dx2[pid] / 2.)
			p.set_velocity(v0[pid] + dv2[pid] / 2.)
		self.compute_forces()
		dx3 = dict( (pid,(vec + dv2[pid] / 2.) * dt) \
		             for pid,vec in v0.iteritems() )
		dv3 = dict( (pid,vec * (dt / pts[pid].weight() ) ) \
		             for pid,vec in self._force.iteritems() )
		
		#K4
		for pid,p in pts.iteritems() :
			p.set_position(x0[pid] + dx3[pid])
			p.set_velocity(v0[pid] + dv3[pid])
		self.compute_forces()
		dx4 = dict( (pid,(vec + dv3[pid]) * dt) \
		             for pid,vec in v0.iteritems() )
		dv4 = dict( (pid,vec * (dt / pts[pid].weight() ) ) \
		             for pid,vec in self._force.iteritems() )
		
		#compute new speed
		for pid,p in pts.iteritems() :
			p.set_velocity(v0[pid] + dv1[pid] / 6. \
			                       + dv2[pid] / 3. \
			                       + dv3[pid] / 3. \
			                       + dv4[pid] / 6.)
		
		#move points
		for pid,p in pts.iteritems() :
			p.set_position(x0[pid] + dx1[pid] / 6. \
			                       + dx2[pid] / 3. \
			                       + dx3[pid] / 3. \
			                       + dx4[pid] / 6.)
		
		#return
		return dt


class RungeKutta2D (RungeKutta) :
	dim = 2

class RungeKutta3D (RungeKutta) :
	dim = 3

########################################################################
#
#	Forward Marching
#
########################################################################
class ForwardMarching (MSSolver) :
	"""Compute equilibrium using a forward marching integration scheme.
	
	At each iteration, compare the forces obtained with the one
	on the previous dt. If force max is bigger, it means that the
	algorihtm will fail to converge and it will use a smaller dt
	"""
	def __init__ (self, point_weight, spring_list, external_constraint,
	                    logout = None) :
		"""Constructor
		
		Initialise integrator.
		
		:Parameters:
		 - `point_weight` (dict of (pid|float) ) - actual mass of points
		 - `spring_list` (list of :class:`Spring`) - elastic interactions
		                                           between points
		 - `external_constraint` (function) - boundary function taking
		                                      this solver as argument
		 - `logout` (outstream) - an outfile stream like structure
		                          used to write some debug informations.
		                          If None, no output messages are printed
		"""
		raise NotImplementedError("#TODO")
		MSSolver.__init__(self,
		                  point_weight,
		                  spring_list,
		                  external_constraint,
		                  logout)
		self._nb_good_iter = 0
	
	def safe_apply_forces (self, dt) :
		ref_pos = dict(self._pos)
		
		while True :
			try :
				self._pos.update(ref_pos)
				self.update_positions(dt)
				return dt
			except DivergenceError :
				dt = dt / 2.
				self._nb_good_iter = 0
				print "div error"
	
	def deform (self, dt) :
		"""Deform positions using ForwardMarching scheme.
		
		compare result with same computation with
		smaller dt
		"""
		pos_prev = dict(self._pos)
		
		Fmax_prev = self.force_max()
		
		if self._nb_good_iter > 10 :
			dt = dt * 1.9
			self._nb_good_iter = 0
			print "debug *2",dt
		
		dt = self.safe_apply_forces(dt)
		self.compute_forces()
		self._nb_good_iter += 1
		while self.force_max() > Fmax_prev :
			print "debug",dt,"%e" % (Fmax_prev - self.force_max())
			self._nb_good_iter = 0
			self._pos.update(pos_prev)
			dt = self.safe_apply_forces(dt / 2.)
			self.compute_forces()
		
		#return
		return dt
	
	def deform_to_equilibrium (self, dt, force_threshold) :
		"""Compute equilibrium positions of the system
		
		While forces are bigger than a given threshold,
		call :func:`deform`
		
		:Parameters:
		 - `dt` (float) - time step
		 - `force_threshold` (float) - small force below
		                  which the system is considered
		                  to be in equilibrium
		"""
		self.compute_forces()
		MSSolver.deform_to_equilibrium(self,dt,force_threshold)

class ForwardMarching2D (ForwardMarching) :
	dim = 2

class ForwardMarching3D (ForwardMarching) :
	dim = 3

File: cmechanics/src/cpp/mass_spring_solver.cpp
/* -*-c++-*- 
 *------------------------------------------------------------------------------
 *                                                                              
 *        openalea.cmechanics: mass spring solver package                       
 *                                                                              
 *        Copyright 2006 INRIA - CIRAD - INRA                      
 *                                                                              
 *        File author(s): Jerome Chopard <jerome.chopard@sophia.inria.fr>         
 *                                                                              
 *        Distributed under the Cecill-C License.                               
 *        See accompanying file LICENSE.txt or copy at                          
 *            http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html       
 *                                                                              
 *        OpenAlea WebSite : http://openalea.gforge.inria.fr                    
 *       
 *                                                                       
 *-----------------------------------------------------------------------------*/

#include <iostream>

#include <algorithm>
#include "cmechanics/mass_spring_solver.h"

namespace mechanics {
	/* *******************************************************
	*
	*	MSSolver2D
	*
	*********************************************************/
	MSSolver2D::~MSSolver2D () {
		/*for (spring_list::iterator it=_spring_list.begin();it!=_spring_list.end();++it) {
			delete *it;
		}*/
		/*for (P2D_list::iterator it=_particules.begin();it!=_particules.end();++it) {
			delete *it;
		}*/
	}

	int MSSolver2D::add_particule () {
		_particules.push_back(Particule() );
		return _particules.size() - 1;
	}
	
	void MSSolver2D::compute_forces () {
		//set forces to zero
		for (P_list::iterator it = _particules.begin();
		                      it != _particules.end();
		                      ++it) {
			it->force = Vector();
		}
		
		//assign forces
		for (spring_list::iterator it = _spring_list.begin();
		                           it != _spring_list.end();
		                           ++it) {
			(*it)->assign_forces(_particules);
		}
		
		//boundary condition
		boundary();
	}
	
	void MSSolver2D::update_positions (double dt) {
		for (P_list::iterator it = _particules.begin();
		                      it != _particules.end();
		                      ++it) {
			it->pos += it->force * (0.5 * dt * dt / it->weight);
			if (it->pos.x() > LARGE or it->pos.y() > LARGE) {
				throw DivergenceError();
			}
		}		
	}
	
	double MSSolver2D::force_max () {
		double Fmax = 0;
		for (P_list::const_iterator it = _particules.begin();
		                            it != _particules.end();
		                            ++it) {
			Fmax = std::max(Fmax,norm( it->force ) );
		}
		return Fmax;
	}
	
	double MSSolver2D::deform (double dt) {
		compute_forces();
		update_positions(dt);
		return dt;
	}
	
	void MSSolver2D::deform_to_equilibrium (double dt, double force_threshold) {
		double Fmax = 2 * force_threshold;
		while ( Fmax > force_threshold ) {
			dt = deform(dt);
			Fmax = force_max();
			std::cout << "Fmax " << Fmax << std::endl;
		}
	}

	/* *******************************************************
	*
	*	MSSolver3D
	*
	*********************************************************/
	MSSolver3D::~MSSolver3D () {
		/*for (spring_list::iterator it=_spring_list.begin();it!=_spring_list.end();++it) {
			delete *it;
		}*/
		/*for (P3D_list::iterator it=_particules.begin();it!=_particules.end();++it) {
			delete *it;
		}*/
	}

	int MSSolver3D::add_particule () {
		_particules.push_back(Particule() );
		return _particules.size() - 1;
	}
	
	void MSSolver3D::compute_forces () {
		//set forces to zero
		for (P_list::iterator it = _particules.begin();
		                      it != _particules.end();
		                      ++it) {
			it->force = Vector();
		}
		
		//assign forces
		for (spring_list::iterator it = _spring_list.begin();
		                           it != _spring_list.end();
		                           ++it) {
			(*it)->assign_forces(_particules);
		}
		
		//boundary condition
		boundary();
	}
	
	void MSSolver3D::update_positions (double dt) {
		for (P_list::iterator it = _particules.begin();
		                      it != _particules.end();
		                      ++it) {
			it->pos += it->force * (0.5 * dt * dt / it->weight);
			if (it->pos.x() > LARGE or it->pos.y() > LARGE or it->pos.z() > LARGE) {
				throw DivergenceError();
			}
		}
	}
	
	double MSSolver3D::force_max () {
		double Fmax = 0;
		for (P_list::const_iterator it = _particules.begin();
		                            it != _particules.end();
		                            ++it) {
			Fmax = std::max(Fmax,norm( it->force ) );
		}
		return Fmax;
	}
	
	double MSSolver3D::deform (double dt) {
		std::cout << "MSSolver" << std::endl;
		compute_forces();
		update_positions(dt);
		return dt;
	}
	
	void MSSolver3D::deform_to_equilibrium (double dt, double force_threshold) {
		double Fmax = 2 * force_threshold;
		while ( Fmax > force_threshold ) {
			dt = deform(dt);
			Fmax = force_max();
			std::cout << "Fmax " << Fmax << std::endl;
		}
	}

	/* *******************************************************
	*
	*	ForwardEuler
	*
	*********************************************************/
	double ForwardEuler2D::deform (double dt) {
		for (int i = 0; i < _nb_steps; ++i) {
			compute_forces();
			update_positions(dt);
		}
		return dt;
	}

	double ForwardEuler3D::deform (double dt) {
		for (int i = 0; i < _nb_steps; ++i) {
			compute_forces();
			update_positions(dt);
		}
		return dt;
	}

	/* *******************************************************
	*
	*	ForwardMarching
	*
	*********************************************************/
	double ForwardMarching2D::deform (double dt) {
		return dt;
	}
	
	double ForwardMarching3D::safe_apply_forces (double dt) {
		P_list p_ref(_particules);
		
		while (true) {
			try {
				copy_positions(p_ref,_particules);
				update_positions(dt);
				return dt;
			}
			catch (DivergenceError) {
				dt /= 2.;
				_nb_good_iter = 0;
			}
		}
	}
	
	double ForwardMarching3D::deform (double dt) {
		//std::cout << "ForwardMarching" << std::endl;
		P_list p_prev(_particules);
		
		double Fmax_prev = force_max();
		
		if (_nb_good_iter > 10.) {
			dt *= 1.9;
			_nb_good_iter = 0;
		}
		
		dt = safe_apply_forces(dt);
		//std::cout << "applyed" << std::endl;
		compute_forces();
		_nb_good_iter += 1;
		while (force_max() > Fmax_prev) {
			//std::cout << "step " << force_max() << "," << Fmax_prev << " dt " << dt << std::endl;
			_nb_good_iter = 0;
			copy_positions(p_prev,_particules);
			dt = safe_apply_forces(dt / 2.);
			compute_forces();
		}
		
		return dt;
	}
	
	void ForwardMarching3D::deform_to_equilibrium (double dt, double force_threshold) {
		//std::cout << "ForwardMarching  equ" << std::endl;
		compute_forces();
		MSSolver3D::deform_to_equilibrium(dt,force_threshold);
	}

//	double ForwardMarching2D::deform (double dt) {
//		compute_forces(particules);
//		return _deform(particules,dt);
//	}
//	
//	double ForwardMarching2D::_deform (P2D_list& particules, double dt) {
//		//compute ref
//		P2D_list p_ref(particules);
//		update_positions(p_ref,dt);
//		
//		//compute test
//		double dt_test = dt / NB_SUBDI;
//		P2D_list p_test(particules);
//		update_positions(p_test,dt_test);
//		
//		for (int i = 0; i < NB_SUBDI; ++i) {
//			compute_forces(p_test);
//			update_positions(p_test,dt_test);
//		}
//		
//		//put forces in particules
//		P2D_list::iterator it = particules.begin();
//		P2D_list::iterator it_test = p_test.begin();
//		while (it != particules.end() and it_test != p_test.end() ) {
//			it->force = it_test->force;
//			++ it;
//			++ it_test;
//		}
//		
//		//compare
//		double err = 0.;
//		P2D_list::iterator it_ref = p_ref.begin();
//		it_test = p_test.begin();
//		while (it_ref != p_ref.end() and it_test != p_test.end() ) {
//			err = std::max(err,norm( it_ref->pos - it_test->pos ) );
//			++ it_ref;
//			++ it_test;
//		}
//		std::cout << "err" << err << "," << dt << std::endl;
//		if (err > ERR_MAX_THRESHOLD) {
//			return _deform(particules,dt / 2.);
//		}
//		else {
//		if (err < ERR_MIN_THRESHOLD) {
//			return std::min(DT_MAX,dt * 1.3);
//		}
//		else {
//			return dt;
//		} }
//	}

//	double ForwardMarching3D::deform (P3D_list& particules, double dt) {
//		//std::cout << "deform" << std::endl;
//		compute_forces(particules);
//		return _deform(particules,dt);
//	}
	
//	double ForwardMarching3D::_deform (P3D_list& particules, double dt) {
//		//std::cout << "_deform" << std::endl;
//		//compute ref
//		//std::cout << "ref" << std::endl;
//		try {
//			P3D_list p_ref(particules);
//			update_positions(p_ref,dt);
//			
//			//compute test
//			//std::cout << "test" << std::endl;
//			double dt_test = dt / NB_SUBDI;
//			P3D_list p_test(particules);
//			update_positions(p_test,dt_test);
//			
//			for (int i = 0; i < NB_SUBDI; ++i) {
//				//std::cout << "test" << i << std::endl;
//				compute_forces(p_test);
//				update_positions(p_test,dt_test);
//			}
//			/*PRINT
//			int i = 0;
//			std::cout << "pos";
//			for (P3D_list::iterator it = p_test.begin();
//			                        it != p_test.end();
//			                        ++it) {
//				std::cout << " " << it->pos;
//				++i;
//			}
//			std::cout << std::endl;*/
//		
//			//put forces in particules
//			P3D_list::iterator it = particules.begin();
//			P3D_list::iterator it_test = p_test.begin();
//			while (it != particules.end() and it_test != p_test.end() ) {
//				it->force = it_test->force;
//				++ it;
//				++ it_test;
//			}
//			
//			//compare
//			double err = 0.;
//			P3D_list::iterator it_ref = p_ref.begin();
//			it_test = p_test.begin();
//			while (it_ref != p_ref.end() and it_test != p_test.end() ) {
//				err = std::max(err,norm( it_ref->pos - it_test->pos ) );
//				++ it_ref;
//				++ it_test;
//			}
//			std::cout << "                     err " << err << "," << prev_error << "," << dt << "," << nb_fine_iter << std::endl;
//			if (prev_error > 0) {
//				//previous step dt *= 1.5
//				if (err > (prev_error * 1.4) and err > ERR_MIN_THRESHOLD) {
//					std::cout << "                     go back" << std::endl;
//					//go back
//					nb_fine_iter = 0;
//					return _deform(particules,dt / 1.5);
//				}
//			}
//			if (err > ERR_MAX_THRESHOLD) {
//				//std::cout << "MAX" << std::endl;
//				prev_error = -1.;
//				nb_fine_iter = 0;
//				return _deform(particules,dt / 2.);
//			}
//			else {
//				//put positions in particules
//				it = particules.begin();
//				it_test = p_test.begin();
//				while (it != particules.end() and it_test != p_test.end() ) {
//					it->pos = it_test->pos;
//					++ it;
//					++ it_test;
//				}
//				if (err < ERR_MIN_THRESHOLD and nb_fine_iter > 10) {
//					prev_error = err;
//					nb_fine_iter = 0;
//					return std::min(DT_MAX,dt * 1.5);
//				}
//				else {
//					prev_error = -1.;
//					nb_fine_iter += 1;
//					return dt;
//				}
//			}
//		}
//		catch (DivergenceError e) {
//			std::cout << "nan " << dt << std::endl;
//			nb_fine_iter = 0;
//			return _deform(particules,dt / 2.);
//		}
//	}

};//mechanics
